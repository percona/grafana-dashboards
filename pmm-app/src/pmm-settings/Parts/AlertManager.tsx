import { PluginTooltip, VerticalFormWrapper } from '../../react-plugins-deps/components/helpers/Helpers';
import { InputField } from '../../react-plugins-deps/components/FieldsComponents/Input/Input';
import { TextAreaField } from '../../react-plugins-deps/components/FieldsComponents/TextArea/TextArea';
import React, { ReactElement, useEffect, useState } from 'react';
import ButtonElement from '../../react-plugins-deps/components/FieldsComponents/Button/Button';
import { Form as FormFinal } from 'react-final-form';
import { useForm } from 'react-final-form-hooks';
import SettingsService from './SettingsService';
import { showErrorNotification, showSuccessNotification } from '../../react-plugins-deps/components/helpers/notification-manager';
import set = Reflect.set;

const AlertManager = props => {
  const [loading, setLoading] = useState(false);

  return (
    <FormFinal
      onSubmit={() => {}}
      render={(): ReactElement => {
        const { form, handleSubmit } = useForm({
          onSubmit: async values => {
            const settings = {
              alert_manager_ip: values.alert_manager_ip,
              alert_manager_rules: values.alert_manager_rules,
            };
            setLoading(true);
            try {
              await SettingsService.setSettings({settings: settings});
              setLoading(false);
              showSuccessNotification({ message: 'Alert manager settings updated' });
            } catch (e) {
              setLoading(false);
              showErrorNotification({ message: e.message });
            }
          },
          validate: () => {},
        });
        useEffect(() => {
          form.initialize(props.settings);
        }, [props.settings]);
        // @ts-ignore
        return (
          <form onSubmit={handleSubmit}>
            <>
              <VerticalFormWrapper
                label={'AlertManager IP'}
                element={<InputField form={form} name={'alert_manager_ip'} placeholder="Enter IP" style={{ width: '100%' }} />}
              />
              <VerticalFormWrapper
                label={'AlertManager rules'}
                element={<TextAreaField form={form} name={'alert_manager_rules'} placeholder="Alert manager rule" style={{ width: '100%' }} />}
                alignLabel={'top'}
              />
              <ButtonElement onClick={() => {}} loading={loading} text={'Apply AlertManager settings'} />
            </>
          </form>
        );
      }}
    />
  );
};

export default AlertManager;
