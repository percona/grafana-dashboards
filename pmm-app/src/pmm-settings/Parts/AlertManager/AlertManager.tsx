import { InputField } from '../../../react-plugins-deps/components/FormComponents/Input/Input';
import { TextAreaField } from '../../../react-plugins-deps/components/FormComponents/TextArea/TextArea';
import React, { ReactElement, useEffect, useState } from 'react';
import ButtonElement from '../../../react-plugins-deps/components/FormComponents/Button/Button';
import { Form as FormFinal } from 'react-final-form';
import SettingsService from '../../Settings.service';
import { showSuccessNotification } from '../../../react-plugins-deps/components/helpers/notification-manager';
import { FormElement } from '../../../react-plugins-deps/components/FormComponents/FormElement/FormElement';
import { PluginTooltip } from '../../../react-plugins-deps/components/helpers/Helpers';
import { css } from 'emotion';
import { GUI_DOC_URL } from '../../panel.constants';

interface AlertManagerSettingsInterface {
  alert_manager_url: string;
  alert_manager_rules: string;
  remove_alert_manager_url?: boolean;
  remove_alert_manager_rules?: boolean;
}

const Styling = {
  textAreaWidth: css`
    width: 100%;
  `,
};

const AlertManager = props => {
  const [loading, setLoading] = useState(false);
  const onSubmit = async (values: AlertManagerSettingsInterface): Promise<void> => {
    const settings: AlertManagerSettingsInterface = {
      alert_manager_url: values.alert_manager_url || '',
      alert_manager_rules: values.alert_manager_rules || '',
    };

    if (!values.alert_manager_url) {
      settings.remove_alert_manager_url = true;
    }

    if (!values.alert_manager_rules) {
      settings.remove_alert_manager_rules = true;
    }

    setLoading(true);
    try {
      await SettingsService.setSettings(settings);
      setLoading(false);
      showSuccessNotification({
        message: 'Alert manager settings updated',
      });
    } catch (e) {
      setLoading(false);
    }
  };
  return (
    <FormFinal
      onSubmit={onSubmit}
      render={({ form, handleSubmit }): ReactElement => {
        useEffect(() => {
          form.initialize(props.settings);
        }, [props.settings]);
        // @ts-ignore
        return (
          <form onSubmit={handleSubmit}>
            <>
              <FormElement
                label="Alertmanager URL"
                tooltip={
                  <PluginTooltip
                    links={[
                      {
                        url: `${GUI_DOC_URL}#prometheus-alertmanager-integration`,
                        text: 'How to integrate Alertmanager with PMM',
                      },
                    ]}
                    text="The URL of the external Alertmanager to use"
                  />
                }
                element={
                  <InputField
                    name="alert_manager_url"
                    placeholder="Enter URL"
                    className={Styling.textAreaWidth}
                  />
                }
              />
              <FormElement
                label="Prometheus Alerting rules"
                tooltip={
                  <PluginTooltip
                    links={[
                      {
                        url: `${GUI_DOC_URL}#prometheus-alertmanager-integration`,
                        text: 'How to integrate Alertmanager with PMM',
                      },
                    ]}
                    text="Alerting rules in the YAML configuration format"
                  />
                }
                element={
                  <TextAreaField
                    name="alert_manager_rules"
                    placeholder="Alerting rules"
                    className={Styling.textAreaWidth}
                  />
                }
                alignLabel="top"
              />
              <ButtonElement loading={loading} text="Apply Alertmanager settings" />
            </>
          </form>
        );
      }}
    />
  );
};

export default AlertManager;
