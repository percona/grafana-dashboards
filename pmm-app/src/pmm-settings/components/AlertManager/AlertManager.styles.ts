import { stylesFactory } from '@grafana/ui';
import { GrafanaTheme } from '@grafana/data';
import { css } from 'emotion';

export const getStyles = stylesFactory((theme: GrafanaTheme) => ({
  alertManagerWrapper: css`
    display: flex;
    flex-direction: column;
  `,
  textarea: css`
    margin: ${theme.spacing.md} 0;
    min-height: 150px;
    width: 600px;
  `,
  input: css`
    margin: ${theme.spacing.md} 0;
  `,
  rulesLabel: css`
    margin-top: ${theme.spacing.sm};
  `,
}));
