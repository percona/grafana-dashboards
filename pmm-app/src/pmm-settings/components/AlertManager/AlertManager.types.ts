import { LoadingCallback } from 'pmm-settings/Settings.service';

export interface AlertManagerProps {
  alertManagerUrl: string;
  alertManagerRules: string;
  updateSettings: (body: any, callback: LoadingCallback) => void;
}
