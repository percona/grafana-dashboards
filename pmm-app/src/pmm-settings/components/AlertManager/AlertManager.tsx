import React, { FC, useState } from 'react';
import { Button, Spinner, useTheme } from '@grafana/ui';
import { cx } from 'emotion';
import { getSettingsStyles } from 'pmm-settings/Settings.styles';
import { Messages } from 'pmm-settings/Settings.messages';
import { LinkTooltip } from 'shared/components/Elements/LinkTooltip/LinkTooltip';
import { getStyles } from './AlertManager.styles';
import { AlertManagerProps } from './AlertManager.types';

export const AlertManager: FC<AlertManagerProps> = ({
  alertManagerUrl,
  alertManagerRules,
  updateSettings
}) => {
  const theme = useTheme();
  const styles = getStyles(theme);
  const settingsStyles = getSettingsStyles(theme);
  const {
    alertmanager: {
      action,
      rulesLabel,
      rulesLink,
      rulesTooltip,
      urlLabel,
      urlLink,
      urlTooltip
    },
    tooltipLinkText
  } = Messages;
  const [url, setUrl] = useState(alertManagerUrl);
  const [rules, setRules] = useState(alertManagerRules);
  const [loading, setLoading] = useState(false);
  const isActionDisabled = () => (
    url === alertManagerUrl
    && rules === alertManagerRules
  );
  const applyChanges = () => {
    const body: any = {
      alert_manager_url: url,
      alert_manager_rules: rules
    };

    if (!url) {
      body.remove_alert_manager_url = true;
    }

    if (!rules) {
      body.remove_alert_manager_rules = true;
    }

    updateSettings(body, setLoading);
  };

  return (
    <div className={styles.alertManagerWrapper}>
      <div
        className={settingsStyles.labelWrapper}
        data-qa="alertmanager-url-label"
      >
        <span>{urlLabel}</span>
        <LinkTooltip
          tooltipText={urlTooltip}
          link={urlLink}
          linkText={tooltipLinkText}
          icon="info-circle"
        />
      </div>
      <input
        className={cx(settingsStyles.input, styles.input)}
        value={url}
        data-qa="alertmanager-url"
        onChange={(e) => setUrl(e.target.value)}
      />
      <div
        className={cx(settingsStyles.labelWrapper, styles.rulesLabel)}
        data-qa="alertmanager-rules-label"
      >
        <span>{rulesLabel}</span>
        <LinkTooltip
          tooltipText={rulesTooltip}
          link={rulesLink}
          linkText={tooltipLinkText}
          icon="info-circle"
        />
      </div>
      <textarea
        className={cx(settingsStyles.textarea, styles.textarea)}
        value={rules}
        data-qa="alertmanager-rules"
        onChange={(e) => setRules(e.target.value)}
      />
      <Button
        className={settingsStyles.actionButton}
        variant="secondary"
        disabled={isActionDisabled() || loading}
        onClick={applyChanges}
        data-qa="alertmanager-button"
      >
        {loading && <Spinner />}
        {action}
      </Button>
    </div>
  );
};
