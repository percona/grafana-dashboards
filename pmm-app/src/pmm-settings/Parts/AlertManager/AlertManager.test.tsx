import React from 'react';
import { shallow } from 'enzyme';
import AlertManager from './AlertManager';

jest.mock('shared/components/helpers/notification-manager');

describe('Settings Part test', () => {
  it('Alert Manager renders correct without props', () => {
    const root = shallow(
      <AlertManager
        settings={{
          alert_manager_url: 'http://localhost',
          alert_manager_rules: 'test rules',
        }}
      />,
    );

    expect(root).toMatchSnapshot();
    root.unmount();
  });
});
