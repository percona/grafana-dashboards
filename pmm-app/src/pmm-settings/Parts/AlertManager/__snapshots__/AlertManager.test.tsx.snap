// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Part test Alert Manager renders correct without props 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="css-150snti"
  >
    <div
      className="ant-row css-15a8aul"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24 css-1ocu2xx"
        style={Object {}}
      >
        <strong>
          Alertmanager URL
        </strong>
        <span>
          <i
            aria-label="icon: question-circle"
            className="anticon anticon-question-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "marginLeft": "5px",
              }
            }
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="question-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0 1 30.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1 0 80 0 40 40 0 1 0-80 0z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
    <div
      className="ant-row css-16lswv3"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24"
        style={Object {}}
      >
        <div
          className="input-field-wrapper"
          style={Object {}}
        >
          <input
            className="input-field input-field--dark css-ybrhvy"
            name="alert_manager_url"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Enter URL"
            style={Object {}}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="ant-row css-16lswv3"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24"
        style={Object {}}
      />
    </div>
  </div>
  <div
    className="css-150snti"
  >
    <div
      className="ant-row css-15a8aul"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24 css-1ocu2xx"
        style={Object {}}
      >
        <strong>
          Prometheus Alerting rules
        </strong>
        <span>
          <i
            aria-label="icon: question-circle"
            className="anticon anticon-question-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "marginLeft": "5px",
              }
            }
          >
            <svg
              aria-hidden="true"
              className=""
              data-icon="question-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0 1 30.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1 0 80 0 40 40 0 1 0-80 0z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
    <div
      className="ant-row css-1q0vyna"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24"
        style={Object {}}
      >
        <div
          className="text-area-field-wrapper"
        >
          <textarea
            className="input-field input-field--textarea input-field--dark css-ybrhvy"
            name="alert_manager_rules"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Alerting rules"
            rows={5}
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="ant-row css-1q0vyna"
      style={Object {}}
    >
      <div
        className="ant-col ant-col-24"
        style={Object {}}
      />
    </div>
  </div>
  <button
    className="ant-btn instance-id-submit-button ant-btn-primary"
    disabled={false}
    onClick={[Function]}
    type="submit"
  >
    <span>
      Apply Alertmanager settings
    </span>
  </button>
</form>
`;
